editor/src/components/canvas/dom-walker.ts

```diff
@@ -307,16 +311,72 @@ function useInvalidateInitCompleteOnMountCount(
 
   return [initCompleteRef.current, setInitComplete]
 }
+type ValueOrUpdater<S> = S | ((prevState: S) => S)
+// todo move to file
+export type SetValueCallback<S> = (
+  valueOrUpdater: ValueOrUpdater<S>,
+  doNotInvalidate?: 'do-not-invalidate',
+) => void
+
+function isSimpleValue<S>(valueOrUpdater: ValueOrUpdater<S>): valueOrUpdater is S {
+  return typeof valueOrUpdater !== 'function'
+}
+function useStateAsyncInvalidate<S>(
+  onInvalidate: () => void,
+  initialState: S,
+): [S, SetValueCallback<S>] {
+  const stateRef = React.useRef(initialState)
+  const setAndMarkInvalidated = React.useCallback(
+    (valueOrUpdater: ValueOrUpdater<S>, doNotInvalidate?: 'do-not-invalidate') => {
+      let resolvedNewValue: S
+      if (isSimpleValue(valueOrUpdater)) {
+        // TODO make this type nicer using a type guard
+        resolvedNewValue = valueOrUpdater
+      } else {
+        resolvedNewValue = valueOrUpdater(stateRef.current)
+      }
+      stateRef.current = resolvedNewValue
+
+      // TODO invalidate
+      if (doNotInvalidate !== 'do-not-invalidate') {
+        onInvalidate()
+      }
+    },
+    [stateRef, onInvalidate],
+  )
+
+  return [stateRef.current, setAndMarkInvalidated]
+}
+
+function useThrottledCallback(callback: () => void): (immediate?: 'immediate') => void {
```

Comment:
I know booleans are bad, but I find `immediate: boolean` much clearer than `immediate: 'immediate' | undefined`.

Reply:
I think 
```typescript
 React.useLayoutEffect(() => {
     fireThrottledCallback('immediate')
   })
```
reads so much better than
```typescript
 React.useLayoutEffect(() => {
     fireThrottledCallback(true)
   })
```

We had / have this big regression in the require function because the createRequireFunction took a boolean parameter and we just missed updating it during a copy-paste, because no one went in to read the type, and the parameter name is not visible on the callsite.